examples/js/loaders/GLTFLoader.js

```diff
@@ -1353,6 +1353,117 @@ THREE.GLTFLoader = ( function () {
+	 *
+	 * @param {Array<Object>} primitives whose length is > 1
+	 * @return {Boolean}
+	 */
+	function isCombinable( primitives ) {
+
+		// Returns true if all primitives use the same attributes/morphAttributes/mode
+		// and also have index. Otherwise returns false.
+
+		var primitive0 = primitives[ 0 ];
+		var targets0 = primitive0.targets || [];
+
+		if ( primitive0.indices === undefined ) return false;
+
+		for ( var i = 1, il = primitives.length; i < il; i ++ ) {
+
+			var primitive = primitives[ i ];
+
+			if ( primitive0.mode !== primitive.mode ) return false;
+			if ( primitive.indices === undefined ) return false;
+
+			if ( Object.keys( primitive0.attributes ).length !== Object.keys( primitive.attributes ).length ) return false;
+
+			for ( var key in primitive0.attributes ) {
+
+				if ( primitive0.attributes[ key ] !== primitive.attributes[ key ] ) return false;
+
+			}
+
+			var targets = primitive.targets || [];
+
+			if ( targets0.length !== targets.length ) return false;
+
+			if ( targets0.length > 0 ) {
+
+				if ( Object.keys( targets0[ 0 ] ).length !== Object.keys( targets[ 0 ] ).length ) return false;
+
+				for ( var j = 0, jl = targets0.length; j < jl; j ++ ) {
+
+					for ( key in targets0[ j ] ) {
+
+						if ( targets0[ j ][ key ] !== targets[ j ][ key ] ) return false;
+
+					}
+
+				}
+
+			}
+
+		}
+
+		return true;
+
+	}
+
+	/**
+	 * Builds a single Mesh with MultiMaterial from meshes.
+	 *
+	 * @param {Array<THREE.Object3D>} meshes whose length is > 1
+	 * @param {Array<Object>} primitives whose length is same as the meshes'
+	 * @param {Array<THREE.Material>} materials
+	 * @return {Object3D}
+	 */
+	function combineMeshes( meshes, primitives, materials ) {
+
+		var newGeometry = meshes[ 0 ].geometry.clone();
+		var newMaterials = [];
+		var indices = []
+		var offset = 0;
+
+		var hasMorphTargets = Object.keys( newGeometry.morphAttributes ).length > 0;
+		var hasMorphNormals = hasMorphTargets && ( newGeometry.morphAttributes.normal !== undefined );
+
+		for ( var i = 0, il = meshes.length; i < il; i ++ ) {
+
+			var mesh = meshes[ i ];
+			var primitive = primitives[ i ];
+
+			var index = mesh.geometry.index;
+			var material = materials[ primitive.material ];
+
+			if ( mesh.isSkinnedMesh ) material.skinning = true;
+			if ( hasMorphTargets ) material.morphTargets = true;
+			if ( hasMorphNormals ) material.morphNormals = true;
+
+			newGeometry.addGroup( offset, index.count, i );
+			newMaterials.push( material );
+
+			for ( var j = 0, jl = index.array.length; j < jl; j ++ ) {
+
+				indices.push( index.array[ j ] );
+
+			}
+
+			offset += index.count;
+
+		}
+
+		newGeometry.setIndex( indices );
+
+		var newMesh = new meshes[ 0 ].constructor( newGeometry, newMaterials );
+
+		newMesh.morphTargetDictionary = Object.assign( {}, meshes[ 0 ].morphTargetDictionary );
+		newMesh.name = meshes[ 0 ].name.slice( 0, meshes[ 0 ].name.lastIndexOf( '_' ) );
+
+		return newMesh;
+
+	}
```

Comment:
Could any of this be a helper method in `BufferGeometryUtils`, instead? Perhaps `BufferGeometryUtils.mergeMeshes( meshes )` and it just handles this case for now, but could potentially handle groups with different start/count properties in the future... I wish we didn't need 120 LOC in GLTFLoader to flatten the scene graph, but I don't see any easier way...

Reply:
Hm, it isn't in core but in `examples/`, is it? I don't really wanna let `GLTFLoader` have any dependencies with other modules in `examples/`.

Reply:
It is in examples, right. We could just skip merging meshes when it isn't available.

I guess my main concerns are that this change

(1) requires a fair bit of code
(2) would be easier to test if it were in a exposed method we can unit test
(3) is specific to a case that only GLTFExporter currently uses; i don't know of any other tools that can create multi-material meshes this way. 

There is already a `BufferGeometryUtils.mergeGeometries` method, so `mergeMeshes` does not seem like a crazy thing to put there, but open to more thoughts on this.

Reply:
Hm, for me, I didn't wanna expose because I thought this'd be kinda specific to glTF (exported by the exporter).

Even if we expose, moving into `BufferGeometryUtils` sounds a bit weird, not crazy tho. If `MeshUtils`, it might be ok. Because I don't wanna BufferGeometry* to think Mesh and Material.